### PR TITLE
feat: freeform content workflow

### PR DESCRIPTION
**Pull request by omBratteng**
WT-1630

### CODE REVIEW

File: Tiltfile [hunk 1]
```diff
@@ -5,7 +5,7 @@ docker_build(
   'api-image',
   context='.',
   dockerfile='./Dockerfile.dev',
-  ignore=['./node_modules', './.infra'],
+  ignore=['./node_modules', './.infra', '__tests__'],
```
Comment (omBratteng) [inline Tiltfile:8]: Not sure if it is needed, it just kept on rebuilding whenever I did a change in tests (often)

File: src/workers/postUpdated.ts [hunk 2]
```diff
@@ -182,22 +197,42 @@ const updatePost = async ({
...
+    ? databasePost.visibleAt ?? data.metadataChangedAt
     : null;
 
+  if (content_type in allowedFieldsMapping) {
+    const allowedFields = [
+      'id',
+      'title',
+      'visible',
+      'visibleAt',
+      'flags',
```
Comment (omBratteng) [inline src/workers/postUpdated.ts:225]: Unsure if `title` should be allowed or not

File: __tests__/workers/postUpdated.ts [hunk 3]
```diff
@@ -237,6 +251,34 @@ it('should update post and modify keywords', async () => {
   expect(postKeywords.length).toEqual(2);
 });
 
+it('should update freeform post and only modify allowed fields', async () => {
+  // await createDefaultKeywords();
```
Comment (rebelchris) [inline __tests__/workers/postUpdated.ts:255]: Probably missed removal :)

File: src/workers/cdc.ts [hunk 4]
```diff
@@ -292,6 +295,19 @@ const onPostChange = async (
     if (data.payload.after.visible) {
       await notifyPostVisible(logger, data.payload.after);
     }
+    if (data.payload.after.type === PostType.Freeform) {
+      const freeform = data as unknown as ChangeMessage<FreeformPost>;
```
Comment (rebelchris) [inline src/workers/cdc.ts:299]: Why is the `as unknown` needed here?

File: src/workers/postUpdated.ts [hunk 5]
```diff
@@ -182,22 +197,42 @@ const updatePost = async ({
...
-    databasePost.metadataChangedAt.toISOString() >= updatedDate.toISOString()
+    databasePost.metadataChangedAt.toISOString() >=
+      data.metadataChangedAt.toISOString()
   ) {
     return null;
   }
 
   const title = data?.title || databasePost.title;
-  const updateBecameVisible = !databasePost.visible && !!title?.length;
+  const updateBecameVisible =
```
Comment (rebelchris) [inline src/workers/postUpdated.ts:207]: Don't think this is particularly right, would mean we always set freeform to true, which might be the case now. but in some instances it might actually be set to false for whatever reason.
I think we can keep title.length, but take the DB value for freeform ones?

File: src/workers/postUpdated.ts (hunk 2 above)
Comment (rebelchris) [reply src/workers/postUpdated.ts:225]: Title should not be allowed to be updated in this flow.

Comment (rebelchris) [reply src/workers/postUpdated.ts:225]: Also technically the ID should never change here.

File: src/workers/postUpdated.ts [hunk 6]
```diff
@@ -291,6 +327,7 @@ const fixData = async ({
   // Try and fix generic data here
   return {
     mergedKeywords,
+    content_type: data?.content_type as PostType,
```
Comment (rebelchris) [inline src/workers/postUpdated.ts:330]: In this flow do you need `content_type` at all?

File: __tests__/workers/postUpdated.ts [hunk 7]
```diff
@@ -237,6 +251,34 @@ it('should update post and modify keywords', async () => {
...
+  });
+  const post = await con.getRepository(ArticlePost).findOneBy({ id: 'p2' });
+  expect(post.metadataChangedAt).toEqual(new Date('2023-01-05T12:00:00.000Z'));
+  expect(post.visible).toEqual(true);
+  expect(post.flags.visible).toEqual(true);
+  expect(post.visibleAt).toEqual(new Date('2023-01-05T12:00:00.000Z'));
+  expect(post.contentCuration).toEqual(['news', 'story', 'release']);
+  expect(post.siteTwitter).toEqual('text');
+
+  expect(post.title).toEqual('test');
```
Comment (rebelchris) [inline __tests__/workers/postUpdated.ts:277]: This should be `freeform post`

**Review by rebelchris — CHANGES_REQUESTED** [review]
Looks great, some minor comments and question.
Overall very neat implementation 👏

File: __tests__/workers/postUpdated.ts (hunk 3 above)
Comment (omBratteng) [reply __tests__/workers/postUpdated.ts:255]: Yeah, removed it now

File: __tests__/workers/postUpdated.ts (hunk 7 above)
Comment (omBratteng) [reply __tests__/workers/postUpdated.ts]: Removed title from allowed fields, and set this to now be equal to `freeform post`.

File: src/workers/postUpdated.ts (hunk 6 above)
Comment (omBratteng) [reply src/workers/postUpdated.ts:329]: Right now only the updatePost uses `content_type`, but I added it to fixData in case createPost would need it in the future.
I can remove it from fixData if that is preferred.

**Comment by omBratteng** [comment]
For some reason I can comment on this comment https://github.com/dailydotdev/daily-api/pull/1399#discussion_r1297179041
> Also technically the ID should never change here.

Yeah. I think I actually could removed the always allowed fields, and just move the entire if section a bit up to line 210-212 someting.

File: src/workers/postUpdated.ts (hunk 2 above)
Comment (omBratteng) [reply src/workers/postUpdated.ts]: Removed the title.

Comment (omBratteng) [reply src/workers/postUpdated.ts]: Yeah, the ID should never change, but to not remove it from the data object, where it is defined a bit further up, I added it to the "always allowed" fields. Same with the rest.
https://github.com/dailydotdev/daily-api/blob/c00f3713688ca1bca848d15bcdbaa2bbc74bfef7/src/workers/postUpdated.ts#L212

File: src/workers/cdc.ts [hunk 8]
```diff
@@ -292,20 +295,50 @@ const onPostChange = async (
...
         if (
           isChanged(freeform.payload.before, freeform.payload.after, 'content')
         ) {
           await notifyPostContentEdited(logger, data.payload.after);
         }
       }
     }
+
+    if (data.payload.after.type === PostType.Freeform) {
+      const freeform = data as unknown as ChangeMessage<FreeformPost>;
```
Comment (rebelchris) [inline src/workers/cdc.ts:327]: you probably also don't need the `as unknown` here.
Really can't recall why it's introduced 😅

File: src/workers/postUpdated.ts (hunk 6 above)
Comment (rebelchris) [reply src/workers/postUpdated.ts:329]: Ah i see, no that's fine with me.
Just didn't see any use so wasn't sure.

Let's keep it for now :)

File: src/workers/postUpdated.ts (hunk 5 above)
Comment (omBratteng) [reply src/workers/postUpdated.ts:207]: Changed it to use the database value.

**Review by rebelchris — APPROVED** [review]
Good from my side.
@idoshamun would love to get your view as well.

**Review by denisb0 — APPROVED** [review]
LGTM
hopefully didn't miss anything important because of limited understanding of the project and language.

**Comment by omBratteng** [comment]
@denisb0 think the most important might be what's getting sent to Yggdrasil
https://github.com/dailydotdev/daily-api/blob/c00f3713688ca1bca848d15bcdbaa2bbc74bfef7/src/workers/cdc.ts#L304-L308

**Comment by denisb0** [comment]
Yes, matches what I am expecting

File: Tiltfile (hunk 1 above)
Comment (idoshamun) [reply Tiltfile:8]: Perfect!

File: src/workers/cdc.ts [hunk 9]
```diff
@@ -292,20 +295,50 @@ const onPostChange = async (
...
+
+    if (data.payload.after.type === PostType.Freeform) {
+      const freeform = data as ChangeMessage<FreeformPost>;
+      if (
+        Math.abs(
+          freeform.payload.before.content.length -
+            freeform.payload.after.content.length,
+        ) >= FREEFORM_POST_MINIMUM_CHANGE_LENGTH
+      ) {
+        await notifyContentRequested(logger, {
```
Comment (idoshamun) [inline src/workers/cdc.ts:334]: Let's abstract this to a function that accepts freeform and triggers the message just in case

File: src/workers/postUpdated.ts [hunk 10]
```diff
@@ -182,22 +197,41 @@ const updatePost = async ({
...
 
   const title = data?.title || databasePost.title;
-  const updateBecameVisible = !databasePost.visible && !!title?.length;
+  const updateBecameVisible =
+    content_type === PostType.Freeform
+      ? databasePost.visible
+      : !databasePost.visible && !!title?.length;
 
   data.id = databasePost.id;
-  data.metadataChangedAt = updatedDate;
```
Comment (idoshamun) [inline src/workers/postUpdated.ts:194]: Why did you remove this line?

Comment (omBratteng) [reply src/workers/postUpdated.ts:194]: I moved it into the fixData function

Comment (omBratteng) [reply src/workers/postUpdated.ts:194]: https://github.com/dailydotdev/daily-api/blob/d614e55bc016f47c98744117e877d52435f7283d/src/workers/postUpdated.ts#L341-L342

File: src/workers/cdc.ts (hunk 9 above)
Comment (omBratteng) [reply src/workers/cdc.ts:334]: Which part?

Comment (idoshamun) [reply src/workers/cdc.ts:334]: The call to notifyContentRequested

Comment (omBratteng) [reply src/workers/cdc.ts]: Extracted the call to a separate function and the if statements

**Review by idoshamun — APPROVED** [review]
